[bug] blank page cached on context canceled

I have a recurring error where the homepage of the website is blank.

When I check the key value in redis, it looks like this:
```
HTTP/0.0 200 OK
Date: Fri, 21 Apr 2023 18:44:19 GMT
X-Souin-Stored-Ttl: 24h0m0s
Content-Length: 0
```

What strikes to me is the key name, which is `GET-https-www.website.com-/`, which all the other keys have `{-VARY-}` appended to them. If I delete the key, and refresh the website, the new key that appears is `GET-https-www.website.com-/{-VARY-}Accept-Encoding:gzip, deflate, br`, and now the page isn't blank anymore.

What would cause that?

```
2023/05/25 19:42:38.124	ERROR	http.log.error	context canceled	{...}, "duration": 0.628182}
panic: Header called after Handler finished

goroutine 56 [running]:
net/http.(*http2responseWriter).Header(0x40003584c0?)
	net/http/h2_bundle.go:6569 +0x80
github.com/darkweak/souin/pkg/middleware.(*CustomWriter).Header(...)
	github.com/darkweak/souin@v1.6.39-0.20230524064538-2cffe5ac9c98/pkg/middleware/writer.go:46
github.com/darkweak/souin/pkg/middleware.(*SouinBaseHandler).Upstream(0x4000990640, 0x40003584c0, 0x4000656c00, 0x40005948e0, 0x4000287c80?, {0x4000144660, 0x22})
	github.com/darkweak/souin@v1.6.39-0.20230524064538-2cffe5ac9c98/pkg/middleware/middleware.go:274 +0x404
github.com/darkweak/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP.func2()
	github.com/darkweak/souin@v1.6.39-0.20230524064538-2cffe5ac9c98/pkg/middleware/middleware.go:478 +0x40
created by github.com/darkweak/souin/pkg/middleware.(*SouinBaseHandler).ServeHTTP
	github.com/darkweak/souin@v1.6.39-0.20230524064538-2cffe5ac9c98/pkg/middleware/middleware.go:477 +0x1324

```

--- a/pkg/middleware/middleware.go
+++ b/pkg/middleware/middleware.go
@@ -378,12 +378,10 @@ func (s *SouinBaseHandler) ServeHTTP(rw http.ResponseWriter, rq *http.Request, n
 	defer s.bufPool.Put(bufPool)
 	customWriter := NewCustomWriter(rq, rw, bufPool)
 	go func(req *http.Request, crw *CustomWriter) {
-		select {
-		case <-req.Context().Done():
-			crw.mutex.Lock()
-			crw.headersSent = true
-			crw.mutex.Unlock()
-		}
+		<-req.Context().Done()
+		crw.mutex.Lock()
+		crw.headersSent = true
+		crw.mutex.Unlock()
 	}(rq, customWriter)
 	s.Configuration.GetLogger().Sugar().Debugf("Request cache-control %+v", requestCc)
 	if !requestCc.NoCache {

--- a/pkg/middleware/middleware.go
+++ b/pkg/middleware/middleware.go
@@ -2,6 +2,7 @@ package middleware
 
 import (
 	"bytes"
+	baseCtx "context"
 	"errors"
 	"fmt"
 	"io"
@@ -279,7 +280,12 @@ func (s *SouinBaseHandler) Upstream(
 		customWriter.Header().Set("Cache-Control", s.DefaultMatchedUrl.DefaultCacheControl)
 	}
 
-	return s.Store(customWriter, rq, requestCc, cachedKey)
+	select {
+	case <-rq.Context().Done():
+		return baseCtx.Canceled
+	default:
+		return s.Store(customWriter, rq, requestCc, cachedKey)
+	}
 }
 
 func (s *SouinBaseHandler) Revalidate(validator *rfc.Revalidator, next handlerFunc, customWriter *CustomWriter, rq *http.Request, requestCc *cacheobject.RequestCacheDirectives, cachedKey string) error {
@@ -388,6 +394,7 @@ func (s *SouinBaseHandler) ServeHTTP(rw http.ResponseWriter, rq *http.Request, n
 			if rfc.ValidateMaxAgeCachedResponse(requestCc, response) != nil {
 				customWriter.Headers = response.Header
 				customWriter.statusCode = response.StatusCode
+				s.Configuration.GetLogger().Sugar().Debugf("Serve from cache %+v", rq)
 				_, _ = io.Copy(customWriter.Buf, response.Body)
 				_, err := customWriter.Send()
 
@@ -466,10 +473,25 @@ func (s *SouinBaseHandler) ServeHTTP(rw http.ResponseWriter, rq *http.Request, n
 		}
 	}
 
-	if err := s.Upstream(customWriter, rq, next, requestCc, cachedKey); err != nil {
-		return err
-	}
+	errorCacheCh := make(chan error)
+	go func() {
+		errorCacheCh <- s.Upstream(customWriter, rq, next, requestCc, cachedKey)
+	}()
 
-	_, _ = customWriter.Send()
-	return nil
+	select {
+	case <-rq.Context().Done():
+		switch rq.Context().Err() {
+		case baseCtx.DeadlineExceeded:
+			customWriter.WriteHeader(http.StatusGatewayTimeout)
+			rw.Header().Set("Cache-Status", cacheName+"; fwd=bypass; detail=DEADLINE-EXCEEDED")
+			_, _ = customWriter.Rw.Write([]byte("Internal server error"))
+			return baseCtx.DeadlineExceeded
+		case baseCtx.Canceled:
+			return baseCtx.Canceled
+		default:
+			return nil
+		}
+	case v := <-errorCacheCh:
+		return v
+	}
 }

--- a/pkg/middleware/middleware.go
+++ b/pkg/middleware/middleware.go
@@ -492,6 +492,9 @@ func (s *SouinBaseHandler) ServeHTTP(rw http.ResponseWriter, rq *http.Request, n
 			return nil
 		}
 	case v := <-errorCacheCh:
+		if v == nil {
+			_, _ = customWriter.Send()
+		}
 		return v
 	}
 }

--- a/pkg/middleware/middleware.go
+++ b/pkg/middleware/middleware.go
@@ -377,6 +377,14 @@ func (s *SouinBaseHandler) ServeHTTP(rw http.ResponseWriter, rq *http.Request, n
 	bufPool.Reset()
 	defer s.bufPool.Put(bufPool)
 	customWriter := NewCustomWriter(rq, rw, bufPool)
+	go func(req *http.Request, crw *CustomWriter) {
+		select {
+		case <-req.Context().Done():
+			crw.mutex.Lock()
+			crw.headersSent = true
+			crw.mutex.Unlock()
+		}
+	}(rq, customWriter)
 	s.Configuration.GetLogger().Sugar().Debugf("Request cache-control %+v", requestCc)
 	if !requestCc.NoCache {
 		validator := rfc.ParseRequest(rq)

--- a/pkg/middleware/writer.go
+++ b/pkg/middleware/writer.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"sync"
 
 	"github.com/darkweak/go-esi/esi"
 	"github.com/darkweak/souin/pkg/rfc"
@@ -24,6 +25,7 @@ func NewCustomWriter(rq *http.Request, rw http.ResponseWriter, b *bytes.Buffer) 
 		Req:        rq,
 		Rw:         rw,
 		Headers:    http.Header{},
+		mutex:      &sync.Mutex{},
 	}
 }
 
@@ -34,12 +36,15 @@ type CustomWriter struct {
 	Req         *http.Request
 	Headers     http.Header
 	headersSent bool
+	mutex       *sync.Mutex
 	statusCode  int
 	// size        int
 }
 
 // Header will write the response headers
 func (r *CustomWriter) Header() http.Header {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
 	if r.headersSent {
 		return http.Header{}
 	}
@@ -48,6 +53,8 @@ func (r *CustomWriter) Header() http.Header {
 
 // WriteHeader will write the response headers
 func (r *CustomWriter) WriteHeader(code int) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
 	if r.headersSent {
 		return
 	}
@@ -76,10 +83,12 @@ func (r *CustomWriter) Send() (int, error) {
 		}
 	}
 
+	r.mutex.Lock()
 	if !r.headersSent {
 		r.Rw.Header().Set("Content-Length", fmt.Sprintf("%d", len(b)))
 		r.Rw.WriteHeader(r.statusCode)
 		r.headersSent = true
 	}
+	r.mutex.Unlock()
 	return r.Rw.Write(b)
 }

--- a/plugins/gin/souin.go
+++ b/plugins/gin/souin.go
@@ -1,7 +1,6 @@
 package gin
 
 import (
-	"bytes"
 	"net/http"
 	"time"
 
@@ -54,7 +53,7 @@ func New(c middleware.BaseConfiguration) *SouinGinMiddleware {
 
 func (s *SouinGinMiddleware) Process() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		_ = s.SouinBaseHandler.ServeHTTP(middleware.NewCustomWriter(c.Request, c.Writer, bytes.NewBuffer([]byte{})), c.Request, func(cw http.ResponseWriter, _ *http.Request) error {
+		_ = s.SouinBaseHandler.ServeHTTP(c.Writer, c.Request, func(cw http.ResponseWriter, _ *http.Request) error {
 			if writer, ok := cw.(gin.ResponseWriter); ok {
 				c.Writer = writer
 			} else if writer, ok := cw.(*middleware.CustomWriter); ok {

